index.js

```diff
@@ -16,58 +14,153 @@ const serializeError = FJS({
   }
 })
 
-function rateLimitPlugin (fastify, opts, next) {
-  const timeWindow = typeof opts.timeWindow === 'string'
-    ? ms(opts.timeWindow)
-    : typeof opts.timeWindow === 'number'
-      ? opts.timeWindow
-      : 1000 * 60
-
-  const store = opts.redis
-    ? new RedisStore(opts.redis, timeWindow)
-    : new LocalStore(timeWindow, opts.cache)
+function buildRouteRate (pluginComponent, params, routeOptions, next) {
+  const urlT = (routeOptions.url === '/') ? 'root' : routeOptions.url.replace(/\//g, ':').slice(1)
 
-  const keyGenerator = typeof opts.keyGenerator === 'function'
-    ? opts.keyGenerator
-    : (req) => req.raw.ip
-
-  const skipOnError = opts.skipOnError === true
-  const max = opts.max || 1000
-  const whitelist = opts.whitelist || []
-  const after = ms(timeWindow, { long: true })
+  const prefix = {
+    cache: params.prefixCache ? params.prefixCache.replace(/[-,.;#*]/g, ':') : urlT
+  }
+  if (pluginComponent.isRedis && pluginComponent.whiteListInRedis) {
+    prefix.wl = urlT
+    pluginComponent.store.addWhiteList(`${pluginComponent.app}:wle:${prefix.wl}`, params.whitelist, (err) => {
+      if (err && params.skipOnError === false) return next(err)
+    })
+  } else {
+    pluginComponent.whitelist.endpoint[prefix.cache] = params.whitelist || []
+  }
 
-  fastify.addHook('onRequest', onRateLimit)
+  routeOptions.preHandler = (req, res, next) => {
+    var key = params.keyGenerator(req)
 
-  function onRateLimit (req, res, next) {
-    var key = keyGenerator(req)
-    if (whitelist.indexOf(key) > -1) {
-      next()
+    if (pluginComponent.isRedis) {
+      if (pluginComponent.whiteListInRedis) {
+        pluginComponent.store.isWhiteList(`${pluginComponent.app}:wlg`, key, function ([err, result]) {
+          if (err && params.skipOnError === false) return next(err)
+          if (result === 1) {
+            next()
+          }
+        })
+        pluginComponent.store.isWhiteList(`${pluginComponent.app}:wle:${prefix.wl}`, key, function ([err, result]) {
+          if (err && params.skipOnError === false) return next(err)
+          if (result === 1) {
+            next()
+          }
+        })
+      }
     } else {
-      store.incr(key, onIncr)
+      if (pluginComponent.whitelist.global.indexOf(key) > -1 || pluginComponent.whitelist.endpoint[prefix.cache].indexOf(key) > -1) {
+        next()
+      }
     }
```

Comment:
I'm not sure about this logic. This PR has already too much change, and it's getting hard to review. Would you mind removing the new whitelist mechanism and send it as a follow-up PR?

Reply:
ok.